library/std/diagnostics.qs

```diff
@@ -31,4 +31,77 @@ namespace Microsoft.Quantum.Diagnostics {
         }
     }
 
+    /// # Summary
+    /// Given two registers, prepares the maximally entangled state
+    /// between each pair of qubits on the respective registers.
+    /// All qubits must start in the |0⟩ state.
+    ///
+    /// # Input
+    /// ## left
+    /// A qubit array in the |0...0⟩ state
+    /// ## right
+    /// A qubit array in the |0...0⟩ state
+    internal operation PrepareEntangledState (left : Qubit[], right : Qubit[]) : Unit
+    is Adj + Ctl {
+        Fact( Length(left) == Length(right), "PrepareEntangledState: Lengths of qubits registers must be equal.");
+        for idxQubit in 0 .. Length(left) - 1  {
+            H(left[idxQubit]);
+            CNOT(left[idxQubit], right[idxQubit]);
+        }
+    }
+    
+    /// # Summary
+    /// Given two operations, asserts that they act identically for all input states.
+    ///
+    /// # Description
+    /// This assertion is implemented by using the Choi–Jamiołkowski isomorphism to reduce
+    /// the assertion to one of a qubit state assertion on two entangled registers.
+    /// Thus, this operation needs only a single call to each operation being tested,
+    /// but requires twice as many qubits to be allocated.
+    /// This assertion can be used to ensure, for instance, that an optimized version of an
+    /// operation acts identically to its naïve implementation, or that an operation
+    /// which acts on a range of non-quantum inputs agrees with known cases.
+    ///
+    /// # Remarks
+    /// This operation requires that the operation modeling the expected behavior is
+    /// adjointable, so that the inverse can be performed on the target register alone.
+    /// Formally, one can specify a transpose operation, which relaxes this requirement,
+    /// but the transpose operation is not in general physically realizable for arbitrary
+    /// quantum operations and thus is not included here as an option.
+    ///
+    /// # Input
+    /// ## nQubits
+    /// Number of qubits to pass to each operation.
+    /// ## actual
+    /// Operation to be tested.
+    /// ## expected
+    /// Operation defining the expected behavior for the operation under test.
+    /// # Output
+    /// True if operations are equal, false otherwise.
+    @Config(Full)
+    operation CheckOperationsAreEqual (
+        nQubits : Int,
+        actual : (Qubit[] => Unit),
+        expected : (Qubit[] => Unit is Adj)) : Bool {
+
+        // Prepare a reference register entangled with the target register.
+        use reference = Qubit[nQubits];
+        use target = Qubit[nQubits];
+
+        // Apply operations.
+        within {
+            PrepareEntangledState(reference, target);
+        } apply {
+            actual(target);
+            Adjoint expected(target);
+        }
+
+        // Check and resturn result.
+        let areEqual = CheckAllZero(reference + target);
+        ResetAll(target);
+        ResetAll(reference);
+        areEqual
```

Comment:
Since all we care about is whether the operations are equal and not which qubit they are unequal for, we don't need 2 * nQubits to do it, just nQubits + 1. This significantly reduces load on the simulator. It also has the nice side benefit of removing the dependency on `PrepareEntangledState` such that the internal operation can be removed.

```suggestion
        // Prepare a reference register entangled with the target register.
        use reference = Qubit();
        use target = Qubit[nQubits];

        // Apply operations.
        within {
            H(reference);
            for qubit in target {
                CNOT(reference, qubit);
            }
        } apply {
            actual(target);
            Adjoint expected(target);
        }

        // Check and resturn result.
        let areEqual = CheckAllZero([reference] + target);
        ResetAll(target);
        Reset(reference);
        areEqual
```

Reply:
Checking Array and the extra qubit separately would probably be faster than concatenating arrays.

Reply:
Yeah, good point... avoids two memory allocations. So something like `CheckAllZero(target) and CheckZero(reference)`.

Reply:
Unfortunately, this won't work. If you take an n-qubit operation, it can be represented by 2ⁿ x 2ⁿ matrix, or 2²ⁿ coefficients. When two n-qubit registers are used as in the original algorithm, it creates a state with 2²ⁿ coefficients. The verification procedure maps matrix coefficients of AB† into the coefficients this state. Therefore, verifying that the matrix is an identity matrix. When we use n-bit register and an additional qubit, we only have 2*2ⁿ coefficients - not sufficient to map and verify all coefficients. So, the verification procedure no longer works.

Here's a short example ([or in playground](https://microsoft.github.io/qsharp/?code=H4sIAAAAAAAACn2QQUvDQBCF7%2FkVc9yAtOrRIlis4CXagl4sPYzpJB0xs3EyQYLkv8vaJG2D%2Bg7LDu%2B9b5YVLKgqMSVImlWNYnWxVJ8rFvAVAQD4kgQSTtVXPrNJl5ksGHPxlXFazf7LzVWx6SI%2Fx82dmDZLz2Iu7puKxl4gQRYXwxU8C1u3P6iuCJQyUpKU4BpW9Subi2cnAUPNyXp3fbk52J9sO5YjYNC9G5BHpKDMK3wECLD02NNu0O3D49MBcbZvjEjtMLWAZfnejDiLuigTTHcs5EbVF7ffvD7f%2FOVcjJ3pFObbt%2FC1YB54S2JszeENw%2B2XvW3URt%2BHc%2ByhDQIAAA%3D%3D)):

```qsharp
namespace MyQuantumProgram {
    open Microsoft.Quantum.Diagnostics;
    open Microsoft.Quantum.Arrays;
    
    @EntryPoint()
    operation Main() : Unit {
        use reference = Qubit();
        use target = Qubit[2];
        within {
            H(reference);
            for qubit in target {
                CNOT(reference, qubit);
            }
        } apply {
            DumpMachine();
            Z(target[0]);
            Z(target[1]);
            // Adjoint to identity
        }
        DumpMachine();
    }
}
```

The state created is only |000⟩+|111⟩ (normalized). Any unitary with eigenvectors |00⟩+|11⟩ will leave this state unchanged. Here we use Z⊗Z. And this procedure cannot distinguish it from the identity gate.

Reply:
Thanks for the detailed explanation and example!